### PR TITLE
Video external check

### PR DESCRIPTION
**Pull request by radutudor**
- check that video is valid by calling the `video-mapper` before scheduling checks
- no more json re-marshalling for perfroming external checks

### CODE REVIEW

File: content/content.go [hunk 1]
```diff
@@ -25,44 +30,95 @@ const logPattern = log.Ldate | log.Ltime | log.Lmicroseconds | log.Lshortfile |
...
+		infoLogger.Printf("External validation for content uuid=[%s] transaction_id=[%s] error: [%v]", p.uuid, p.txID, string(bs))
+	}
+
+	return statusCheck(resp.StatusCode)
+}
+
+func cleanupResp(resp *http.Response) {
+	_, err := io.Copy(ioutil.Discard, resp.Body)
+	if err != nil {
+		warnLogger.Printf("[%v]", err)
```
Comment (bogdanguranda) [inline content/content.go]: I think this logging message should have more a bit, like: "Validation clean up process failed, error was: %v", err
Same for the logging in the if statement below.

File: content/content.go [hunk 2]
```diff
@@ -25,44 +30,95 @@ const logPattern = log.Ldate | log.Ltime | log.Lmicroseconds | log.Lshortfile |
...
+		return false
+	}
+
+	resp, err := httpCaller.DoCallWithEntity(
+		"POST", p.validationURL, p.username, p.password,
+		checks.ConstructPamTxId(p.txID),
+		"application/json", bytes.NewReader(p.binaryContent))
+
+	if err != nil {
+		warnLogger.Printf("External validation for content uuid=[%s] transaction_id=[%s] error: [%v]. Skipping external validation.", p.uuid, p.txID, err)
```
Comment (bogdanguranda) [inline content/content.go]: On line 225 of app.go, almost the same message is logged (when getting false from IsValid, which is coming up from this method). It looks kind of duplicate logging, but maybe it's fine to keep them both, but also the one in app.go logs as INFO and this one as WARN, at least we should log the same level, what do you think?

Comment (radutudor) [inline content/content.go]: I think it's different. In `app.go` you are checking the boolean return of a method, while here you have encountered an actual error, which you are logging.

Btw, I don't really understand what could lead to an error taking place during `httpCaller.DoCallWithEntity`. If anything is not configured properly, I would expect getting a http error status in the response...

File: content/content.go (hunk 1 above)
Comment (radutudor) [inline content/content.go]: Will do.

File: content/video.go [hunk 3]
```diff
@@ -43,3 +64,11 @@ func (v Video) GetType() string {
 func (v Video) GetUUID() string {
 	return v.UUID
 }
+
+func videoStatusCheck(status int) bool {
```
Comment (radutudor) [inline content/video.go]: right 👍

File: content/video_test.go [hunk 4]
```diff
@@ -1,69 +1,118 @@
...
+		defer req.Body.Close()
+		reqBody, err := ioutil.ReadAll(req.Body)
+		assert.NoError(t, err)
+		assert.Equal(t, videoInvalid.BinaryContent, reqBody)
+
+		w.WriteHeader(http.StatusBadRequest)
+	}))
+
+	if videoInvalid.IsValid(testServer.URL+"/map", txId, "", "") {
+		t.Error("Video should fail extrnal validation.")
```
Comment (radutudor) [inline content/video_test.go]: :(

File: content/content.go (hunk 2 above)
Comment (bogdanguranda) [inline content/content.go]: Ok, right, so the error would be some third-party library failure, which is really an unexpected case of error, and should be treated differently, I agree.

**Comment by kovacshuni** [comment]
this was worth it! 👍

File: content/content.go [hunk 5]
```diff
@@ -29,44 +34,95 @@ const logPattern = log.Ldate | log.Ltime | log.Lmicroseconds | log.Lshortfile |
...
+		return ValidationResponse{false, deletedCheck()}
+	}
+
+	resp, err := httpCaller.DoCallWithEntity(
+		"POST", p.validationURL, p.username, p.password,
+		checks.ConstructPamTxId(p.txID),
+		"application/json", bytes.NewReader(p.binaryContent))
+
+	if err != nil {
+		warnLogger.Printf("External validation for content uuid=[%s] transaction_id=[%s] error: [%video]. Skipping external validation.", p.uuid, p.txID, err)
```
Comment (tosan88) [inline content/content.go]: What would you obtain with `%video`? I guess you would like only `%v`: https://golang.org/pkg/fmt/#hdr-Printing

Comment (tosan88) [inline content/content.go]: Also applies for other places as well ^^

**Comment by tosan88** [comment]
I'm sorry I didn't see this before, but can you please coordinate with us (Video Team) before releasing this? It might trip on us otherwise. Thanks!

File: content/content.go (hunk 5 above)
Comment (radutudor) [inline content/content.go]: I refactored to rename all `Video` type receivers from `v` to `video` and Idea was a bit overeager and also replaced all text occurrences (including the the fmt strings)

Thanks for noticing this and for the review! Will fix.

**Comment by radutudor** [comment]
At the moment this PR is not even in testing. When we are closer to release, we'll sync up so that we don't mess you guys up...

**Comment by bogdanguranda** [comment]
This was merge in to this PR: https://github.com/Financial-Times/publish-availability-monitor/pull/135
Release for this will be made along with the changes there.

**Pull request closed without merging**